content/influxdb/cloud-serverless/process-data/downsample_quix.md

```diff
@@ -0,0 +1,333 @@
+
+if localdev == "false":
+    # Create a Quix platform-specific application instead (broker address is in-built)
+    app = Application.Quix(consumer_group=consumer_group_name, auto_create_topics=True)
+
+input_topic = app.topic('downsampled-data', value_deserializer=JSONDeserializer())
+
+## ... remaining code trunctated for brevity ...
+
+def send_data_to_influx(message):
+    logger.info(f"Processing message: {message}")
+    try:
+
+        ## ... remaining code trunctated for brevity ...
+
+        # Construct the points dictionary
+        points = {
+            "measurement": measurement_name,
+            "tags": tags,
+            "fields": fields,
+            "time": message['time']
+        }
+
+        influxdb_downsampled.write(record=points, write_precision="ms")
+
+sdf = app.dataframe(input_topic)
+sdf = sdf.update(send_data_to_influx) # Continuously apply the 'send_data' function to each message in the incoming stream
+
+## ... remaining code trunctated for brevity ...
+```
+{{% /code-placeholders %}}
+
+* You can find the full code for this process in the [Quix GitHub repository](https://github.com/quixio/template-influxdbv3-downsampling/blob/dev/InfluxDB%20V3%20Data%20Sink/main.py).
+
+## Get the full downsampling code files
+
+To get the complete set of files referenced in this tutorial, you can clone the Quix "downsampling template" repository or use an interactive version of this tutorial saved as a Jupyter Notebook.
+
+### Clone the downsampling template repository
+
+To clone the downsampling template, enter the following command in the command line:
+
+`git clone https://github.com/quixio/template-influxdbv3-downsampling.git`
+
+This repository contains the following folders which store different parts of the whole pipeline:
+
+
+* **Machine Data to InfluxDB**- A script that generates synthetic machine data and writes it to InfluxDB (useful if you dont have your own data yet, or just want to work with test data first).
+   * It produces a reading every 250 milliseconds. 
+   * This script originally comes from the [InfluxCommunity repository](https://github.com/InfluxCommunity/Arrow-Task-Engine/blob/master/machine_simulator/src/machine_generator.py) but has been adapted to write directly to InfluxDB rather than using an MQTT broker.
+* **InfluxDB V3 Data Source** - A service that queries for fresh data from InfluxDB at specific intervals. It's configured to look for the measurement produced by the previously-mentioned synthetic machine data generator. It writes the raw data to a Kafka topic called "raw-data".
+* **Downsampler** - A service that performs a 1-minute tumbling window operation on the data from InfluxDB and emits the mean of the "temperature" reading every minute. It writes the output to a "downsampled-data" Kafka topic.
+* **InfluxDB V3 Data Sink** - A service that reads from the "downsampled-data" topic and writes the downsample records as points back into InfluxDB.
```

Comment:
```suggestion
- **Machine Data to InfluxDB**: A script that generates synthetic machine data
  and writes it to InfluxDB. This is useful if you dont have your own data yet,
  or just want to work with test data first.

  - It produces a reading every 250 milliseconds. 
  - This script originally comes from the
    [InfluxCommunity repository](https://github.com/InfluxCommunity/Arrow-Task-Engine/blob/master/machine_simulator/src/machine_generator.py)
    but has been adapted to write directly to InfluxDB rather than using an MQTT broker.
    
- **InfluxDB V3 Data Source**: A service that queries for fresh data from
  InfluxDB at specific intervals. It's configured to look for the measurement
  produced by the previously-mentioned synthetic machine data generator.
  It writes the raw data to a Kafka topic called "raw-data".
- **Downsampler**: A service that performs a 1-minute tumbling window operation
  on the data from InfluxDB and emits the mean of the "temperature" reading
  every minute. It writes the output to a "downsampled-data" Kafka topic.
- **InfluxDB V3 Data Sink**: A service that reads from the "downsampled-data"
  topic and writes the downsample records as points back into InfluxDB.
```